api/http/langos/langos.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package langos
+
+import (
+	"io"
+	"sync"
+)
+
+// Reader contains all methods that Langos needs to read data from.
+type Reader interface {
+	io.ReadSeeker
+	io.ReaderAt
+}
+
+// Langos is a reader with a lookahead peekBuffer
+// this is the most naive implementation of a lookahead peekBuffer
+// it should issue a lookahead Read when a Read is called, hence
+// the name - langos
+// |--->====>>------------|
+//    cur   topmost
+// the first read is not a lookahead but the rest are
+// so, it could be that a lookahead read might need to wait for a previous read to finish
+// due to resource pooling
+//
+// All Read and Seek method call must be synchronous.
+type Langos struct {
+	reader    Reader // reader needs to implement io.ReadSeeker and io.ReaderAt interfaces
+	size      int64
+	cursor    int64 // current read position
+	peeks     []*peek
+	peekSize  int
+	closed    chan struct{} // terminates peek goroutine and unblocks Read method
+	closeOnce sync.Once     // protects closed channel on multiple calls to Close method
+}
+
+// NewLangos bakes a new yummy langos that peeks
+// on provided reader when its Read method is called.
+// Argument peekSize defines the length of peeks.
+func NewLangos(r Reader, peekSize int) *Langos {
```

Comment:
The package is called `langos` and this constructor is also called `NewLangos`, so ultimately we end up with `langos.NewLangos` which is not very user-friendly unless you are intrinsically familiar with this code. `LookaheadBufferedReadSeeker` or smth like that would be much better than `langos` imo.

Reply:
Yes, Langos is used as a codename and is not so descriptive. We can rename it, but it would be nice to have something shorter. Since this is the only codename in the package with the same name as the package. Maybe a package doc comment that describes what langos stands for would help.